accepted/2024/tensors/README.md

```diff
@@ -0,0 +1,424 @@
+
+## Designs
+
+The overall design has been influenced by the existing .NET surface area and the generic math feature (https://learn.microsoft.com/en-us/dotnet/standard/generics/math).
+
+### Base Interface
+
+At it's core there is an `ITensor<TSelf, T>` interface that forms the basis of interchange between types. This interface defines the minimal surface area expected for users to be able interact with a tensor.
+
+Different libraries support different sets of `T`. For example, some only support numeric `T`, some also support `string` or `bool`, and some even support more esoteric concepts like `DateTime`. .NET does not, today, have a system where we can say that members are available when a `T` meets a given constraint and so, much like with `generic math`, the number of interfaces we expose here "could" become quite expansive (potentially mirroring the generic math interface hierarchy). Extension methods do largely work, but they aren't possible today for `properties` or `operators` and so this somewhat limits our choices. However, `extension everything` has been a long standing request and this might be the right time to push for the support as it greatly simplifies our surface area here.
+
+To keep things simple, and in hopes of some support for extension `properties` and `operators`, this initially just defines the core tensor type which allows access to the data following a particular shape, to take slices of that data, and to reshape that data.
+
+```csharp
+namespace System.Numerics.Tensors;
+
+public interface ITensor<TSelf, T>
+    : IEnumerable<T>,
+      IEquatable<TSelf>,
+      IEqualityOperators<TSelf, TSelf, bool>
+    where TSelf : ITensor<TSelf, T>
+    where T : IEquatable<T>, IEqualityOperators<T, T, bool>
+{
+    // TODO: Determine if we can implement `IEqualityOperators<TSelf, T, bool>`.
+    // It looks like C#/.NET currently hits limitations here as it believes TSelf and T could be the same type
+    // Ideally we could annotate it such that they cannot be the same type and no conflicts would exist
+
+    static TSelf Empty { get; }
+
+    bool IsEmpty { get; }
```

Comment:
How is empty defined? If I have a two-dimensional tensor where the length of both dimensions is 0, I assume that's empty?  Do we need to an empty singleton for different numbers of dimensions, or just having whatever shape Empty returns is fine?

Reply:
> How is empty defined? 

All lengths are 0, just like for array.

> Do we need to an empty singleton for different numbers of dimensions, or just having whatever shape Empty returns is fine?

For empty in particular, I don't think we need one per `rank`. Rather, it can be a special marker value that is treated as compatible with other sizes, much as `scalar` is implicitly broadcast to every element for many of the operations.